misc/python/materialize/cli/run.py

```diff
@@ -184,7 +185,36 @@ def main() -> int:
         raise UIError(f"unknown program {args.program}")
 
     print(f"$ {' '.join(command)}")
-    os.execvp(command[0], command)
+    child_pid = os.fork()
+    assert child_pid >= 0
+    if child_pid == 0:
+        try:
+            os.setsid()
+            os.setpgid(os.getpid(), os.getpid())
+        except OSError:
+            pass
+        _set_foreground_process(os.getpid())
+        os.execvp(command[0], command)
+
+    try:
+        os.setpgid(child_pid, child_pid)
+    except OSError:
+        pass
+    _set_foreground_process(child_pid)
+    (_, ws) = os.wait()
+    try:
+        os.killpg(child_pid, signal.SIGTERM)
+    except ProcessLookupError:
+        pass
+    exit(os.waitstatus_to_exitcode(ws))
+
+
+def _set_foreground_process(pid: int) -> None:
+    if not os.isatty(2):
+        return
+    signal.signal(signal.SIGTTOU, signal.SIG_IGN)
+    with open(os.ttyname(2), "w") as tty:
+        os.tcsetpgrp(tty.fileno(), os.getpgrp())
```

Comment:
It's been a long time since I've thought about Unix signals and process groups, but is it strictly necessary to foreground the child and move the parent into the child's process group? I would have expected that it'd be sufficient to have the child create its own process group before exec'ing, so that the parent could unconditionally SIGTERM that process group before it exits. But that the parent could stay as the foreground process in the terminal. Maybe that breaks how signals are delivered to `environmentd` or `sqllogictest`? But then again neither of those processes have any special signal handling, nor do they expect to read input, or anything like that.

Reply:
If you want to ^C sqllogictest (or environmentd) and actually have that process die, it needs to be in the foreground.  Since my usual workflow is to run envd, do some things, and then ^C it, this is necessary.  This wasn't the case before because we exec'd the binary.  It wasn't necessary in the shell case because trap does a lot for us -- doing this is actually shorter than getting the signal handling right to emulate trap exit.

Reply:
By the way, we don't move the parent into the child's process group.  What's going on there is what I allude to in the commit message, an unfortunate historical race condition.  (See the source to bash or dash, around their setpgrp/tcsetpgrp calls, for examples and discussion.)

Reply:
> What's going on there is what I allude to in the commit message, an unfortunate historical race condition, where both the parent and the child have to try to put the child into its own process group. (See the source to bash or dash, around their setpgrp/tcsetpgrp calls, for examples and discussion.)

Oh. I see.